staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -728,9 +744,10 @@ func shouldDelegateList(opts storage.ListOptions) bool {
 	pred := opts.Predicate
 	match := opts.ResourceVersionMatch
 	consistentListFromCacheEnabled := utilfeature.DefaultFeatureGate.Enabled(features.ConsistentListFromCache)
+	requestWatchProgressSupported, _ := etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
```

Comment:
There are places where we use `cacher.Supports` and there are places where we're using `DefaultFeatureSuportChecker.Supports` - this is misleading.

We should decide on a single way of doing that - though I'm still thinking which one:
1) using global variables isn't good, which suggests cacher.Supports
2) otoh, we're using this global variable underneath anyway....
3) here we want consistency with the other `shouldDelegate` function, and given we use DefaultFeatureSupportChecker here, maybe we can just use it in other two places too and avoid extending the storage interface at all?

@serathius

Reply:
There are 3 flows in the code:
* Bootstrap, when we initialize storage we call `CheckClient` - Needs to  access to both storage and client.
* When calculating APF for list request we call `Supports` - no access to storage, needs a global variable. 
* When serving LIST request we call both `CheckClient` and `Supports` - access to storage.

The standout is the APF calculation not having access to storage, thus needing a global variable. As we discussed before, we should refactor the APF code to be aware of the storage. That would address the need for a public variable and allow us to make `DefaultFeatureSuportChecker` private to `etcd3` package.

As for alternatives you described, the second one about using global variable, we need to have a global variable to maintain a per etcd endpoint cache. However, using it requires access to etcd client. This client can have different per resource configuration, so we cannot reuse it between clients. This would mean that that storage interface would need to extended to expose client somehow, breaking it's abstraction. E.g. `DefaultFeatureSuportChecker.CheckForClient(storage.ClientConfig())`

In the third approach, we cannot reuse `shouldDelegate` results from checking APF to serving LIST, the reason is fact that we need to call `CheckClient`. If apiserver starts before etcd, etcd might be down and we will not get any response from it about it's version. We need to continue checking etcd endpoints during LIST.

Reply:
If we decide to keep the `cacher.Supports` as I suggested, we can either:
* Merge the PR as is with `DefaultFeatureSuportChecker` public temporarily until we refactor APF.
* Block the PR on refactoring APF

Reply:
Let's not block on refactoring APF, this will be a much bigger change and I'm not convinced it will land in 1.31 timeframe.

But I guess my point was different. Given that we need a `DefaultFeatureSupportChecker` in APF code (and probably here for consistency too), shouldn't we just use it also in the other two places instead of extending the storage.Interface?
In other words, e.g. in line 453 above switch to:
```
supported, err := etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
```
and get rid of extending storage.Interface with Supports function.
This makes the change simpler and more consistent.

And once this is merged, the follow-up is to refactor APF code and based on that change the storage.Interface and make defaultfeaturesupportChecker private.

Reply:
> But I guess my point was different. Given that we need a `DefaultFeatureSupportChecker` in APF code (and probably here for consistency too), shouldn't we just use it also in the other two places instead of extending the storage.Interface? In other words, e.g. in line 453 above switch to:
> 
> ```
> supported, err := etcdfeature.DefaultFeatureSupportChecker.Supports(storage.RequestWatchProgress)
> ```
> 
> and get rid of extending storage.Interface with Supports function. This makes the change simpler and more consistent.

We can call `etcdfeature.DefaultFeatureSupportChecker.Supports` directly in APF, however it's not so simple for calling `etcdfeature.DefaultFeatureSupportChecker.CheckForClient` when serving LIST request. We need to access client and it's endpoints somehow. This is not exposed by `storage` interface and only available within etcd3 implementation. `storage.Supports` calls both `DefaultFeatureSupportChecker.CheckForClient` and `DefaultFeatureSupportChecker.Supports`. It mostly functions as a wrapper for `CheckForClient` that provides the client.

Reply:
So we're boiling down on sth that I don't like, which is calling `CheckForClient` as part of `storage.Support`.
And IIUC, this is effectively done to ensure that checker will be properly initialized but at the same time handle the cases where kube-apiserver starts before etcd.

So I would like to revisit this pattern, what if instead (which is touching on your other comment too) we would do the following:
- as part of Cacher initialization, or actually even better as part of etcd3.Store creation we would start a goroutine, that will be periodically calling `CheckForClient` until it succeeds [will not get error]

That gives us a cleaner design where we don't call CheckForClient implicitly (which is strange) - and we also don't need the `storage.Supports()` method at all.

Reply:
Makes sense to me, it would also solve the bootstrap issue from my other comment https://github.com/kubernetes/kubernetes/pull/124612#discussion_r1584323649.

cc @ah8ad3  that will require some redesign in DefaultFeatureSupportChecker so that we create only one goroutine per etcd endpoint.

Reply:
Thanks @wojtek-t @serathius, that's a good solution. Let me see what can i do and will get in touch.